### PR TITLE
Refactoring: boot.exe on windows and cache for installed tools

### LINKED ISSUE #6: Are downloads cached?
Hi, I’m interested in using this action, but I can’t tell whether it caches Clojure and whichever tools are specified (in my case it’d be tools.deps along with the CLI tools.) I’m generally uncomfortable with automations that download the same dependencies every time they run, so I personally tend to include very comprehensive caching in my automations ([example](https://github.com/FundingCircle/md2c8e/pull/26)).

I did search the code and saw that it’s using `@actions/tool-cache` which seems promising, but when I read [the docs](https://github.com/actions/toolkit/tree/master/packages/tool-cache#cache) I found this:

> Finally, you can cache these directories in our tool-cache. This is useful if you want to switch back and forth between versions of a tool, or save a tool between runs **for self-hosted runners**.

(emphasis mine)

This seems to imply that `@actions/tool-cache` does not actually, you know, _cache_ anything, when running on GitHub’s runners. But of course, there’s a very good chance I’m just missing something here.

Bottom line: does this action currently its downloads when running on GitHub’s runners?

Thank you!

### LINKED ISSUE #57: boot support for MS-Windows
Hi,

is there a technical reason why `boot` is not supported for MS-Windows? `boot.exe` appears to behave similar to `boot.sh`, both available from the same source at https://github.com/boot-clj/boot-bin/releases.

```yml
      - name: Get boot version
        # Boot is not yet available for windows
        if: ${{ matrix.os != 'windows-latest' }}
        run: boot -V
```

Thanks

### PR DESCRIPTION
**Pull request by DeLaGuardo**
closes #57
closes #6 

Added pre and post execution steps to cache installed tools.

### CODE REVIEW

File: src/boot.ts [hunk 1]
```diff
@@ -100,10 +108,22 @@ async function installBoot(
       env['JAVA_CMD'] = process.env['JAVA_CMD']
     }
 
-    await exec.exec(`./boot ${version === 'latest' ? '-u' : '-V'}`, [], {
-      cwd: path.join(destinationFolder, 'boot', 'bin'),
-      env
-    })
+    if (utils.isWindows()) {
+      await exec.exec(
+        `reg add "HKLM\\SOFTWARE\\JavaSoft\\Java Development Kit\\1.8" /v JAVAHome /d "$env:JAVA_HOME" /f`
```
Comment (zharinov) [inline src/boot.ts:113]: Hopefully, Java versions can be quite safely extracted from `JAVA_HOME`

Comment (DeLaGuardo) [reply src/boot.ts:113]: I count on that as well, but right now I tried to make it work with at least one known JDK but without success

Comment (DeLaGuardo) [reply src/boot.ts]: Most likely, this message coming from Launch4j where the minimum JRE is set to 1.7.0, so everything above that should work

Comment (DeLaGuardo) [reply src/boot.ts]: ok, it works, but for some reason, the workaround must be a part of the workflow. I'll try it this evening to check what is missing when using `exec(...)` inside of the action.

**Comment by zharinov** [comment]
FYI most of pre/post handler usages I've found prefer to use same entrypoint and dispatch based on current stage. The alternative is figuring out how to build 3 different `js` files as entrypoints.

**Comment by DeLaGuardo** [comment]
> FYI most of pre/post handler usages I've found prefer to use same entrypoint and dispatch based on current stage. The alternative is figuring out how to build 3 different `js` files as entrypoints.
> 
> UPD. Though your approach looks also nice, here is my take: https://github.com/zharinov/setup-clojure/blob/test-pre-post/src/entrypoint.ts

Nice! It looks much better than building separated scripts. I'll give it a try.

**Pull request merged**